hudi-spark-datasource/hudi-spark/src/test/java/org/apache/hudi/TestHoodiePositionBasedFileGroupRecordBuffer.java

```diff
@@ -0,0 +1,214 @@
+import static org.apache.hudi.common.testutils.RawTripTestPayload.recordsToStrings;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class TestHoodiePositionBasedFileGroupRecordBuffer extends TestHoodieFileGroupReaderOnSpark {
+  private final HoodieTestDataGenerator dataGen = new HoodieTestDataGenerator(0xDEEF);
+  private HoodieTableMetaClient metaClient;
+  private Schema avroSchema;
+  private HoodiePositionBasedFileGroupRecordBuffer<InternalRow> buffer;
+  private String partitionPath;
+
+  public void prepareBuffer(boolean useCustomMerger) throws Exception {
+    Map<String, String> writeConfigs = new HashMap<>();
+    writeConfigs.put(HoodieStorageConfig.LOGFILE_DATA_BLOCK_FORMAT.key(), "parquet");
+    writeConfigs.put(KeyGeneratorOptions.RECORDKEY_FIELD_NAME.key(), "_row_key");
+    writeConfigs.put(KeyGeneratorOptions.PARTITIONPATH_FIELD_NAME.key(), "partition_path");
+    writeConfigs.put("hoodie.datasource.write.precombine.field", "timestamp");
+    writeConfigs.put("hoodie.payload.ordering.field", "timestamp");
+    writeConfigs.put(HoodieTableConfig.HOODIE_TABLE_NAME_KEY, "hoodie_test");
+    writeConfigs.put("hoodie.insert.shuffle.parallelism", "4");
+    writeConfigs.put("hoodie.upsert.shuffle.parallelism", "4");
+    writeConfigs.put("hoodie.bulkinsert.shuffle.parallelism", "2");
+    writeConfigs.put("hoodie.delete.shuffle.parallelism", "1");
+    writeConfigs.put("hoodie.merge.small.file.group.candidates.limit", "0");
+    writeConfigs.put("hoodie.compact.inline", "false");
+    writeConfigs.put(HoodieWriteConfig.WRITE_RECORD_POSITIONS.key(), "true");
+    commitToTable(recordsToStrings(dataGen.generateInserts("001", 100)), INSERT.value(), writeConfigs);
+
+    String[] partitionPaths = dataGen.getPartitionPaths();
+    String[] partitionValues = new String[1];
+    partitionPath = partitionPaths[0];
+    partitionValues[0] = partitionPath;
+
+    metaClient = createMetaClient(getStorageConf(), getBasePath());
+    avroSchema = new TableSchemaResolver(metaClient).getTableAvroSchema();
+    Option<String[]> partitionFields = metaClient.getTableConfig().getPartitionFields();
+    Option<String> partitionNameOpt = StringUtils.isNullOrEmpty(partitionPaths[0])
+        ? Option.empty() : Option.of(partitionPaths[0]);
+
+    HoodieReaderContext ctx = getHoodieReaderContext(getBasePath(), avroSchema, getStorageConf());
+    ctx.setTablePath(metaClient.getBasePathV2().toString());
+    ctx.setLatestCommitTime(metaClient.createNewInstantTime());
+    ctx.setShouldMergeUseRecordPosition(true);
+    ctx.setHasBootstrapBaseFile(false);
+    ctx.setHasLogFiles(true);
+    ctx.setNeedsBootstrapMerge(false);
```

Comment:
So `hasBootstrapBaseFile` as true and `needsBootstrapMerge` as true are covered by other tests, correct?

Reply:
Yes covered in TestFiltersInFileGroupReader and also some existing bootstrap tests should cover it